File: challenges/1544/js/solution.ts
import { expect } from "../../util";

const f = (s: string): string => {
  if (!s.length) {
    return s;
  }

  const stack = [];
  for (let i = 0; i < s.length; i++) {
    const topChar = stack.length ? stack[stack.length - 1] : "";
    if (
      topChar &&
      topChar !== s[i] &&
      topChar.toLowerCase() === s[i].toLowerCase()
    ) {
      stack.pop();
    } else {
      stack.push(s.charAt(i));
    }
  }

  return stack.join("");
};

expect(f(""), "");
expect(f("leEeetcode"), "leetcode");
expect(f("abBAcC"), "");

File: challenges/1544/python/solution.py
#!/usr/bin/env python

from challenges.utils import expect


def f(s: str) -> str:
    if not s:
        return s

    stack = [s[0]]
    for i in range(1, len(s)):
        if stack and abs(ord(s[i]) - ord(stack[-1])) == 32:
            stack.pop(-1)
        else:
            stack.append(s[i])

    return "".join(stack)


if __name__ == "__main__":
    expect(f(""), "")
    expect(f("leEeetcode"), "leetcode")
    expect(f("abBAcC"), "")
